Make `jitsi_server_url` an organization-level setting.

HemantUmre12 opened the pull request:
<!-- Describe your pull request here.-->

Fixes: #17914 

Related PR: #18455, #19788

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Screenshots and screen captures:</summary>

![image](https://github.com/zulip/zulip/assets/87542880/c10d51d6-742b-4d72-b868-b5e89fd99545)

![image](https://github.com/zulip/zulip/assets/87542880/1f540d52-fe72-4d86-87b7-8f5c15f7242f)

![image](https://github.com/zulip/zulip/assets/87542880/3a8e9d87-ebde-4418-8146-0e6a34838428)

![image](https://github.com/zulip/zulip/assets/87542880/ab53c9f8-d6e6-4750-8661-22ef428bf600)

</details>

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 reviewed (commented):
Posted few comments about naming.

Also a couple of UX issues -
- The server level url can be long, so we need to handle the select dropdown and the custom input in such cases. Currently only partial url is visible in the dropdown.
- We can also disable the `Save changes` button till the input in custom input is not valid like we do for custom inputs for time limit settings.

HemantUmre12 commented:
@sahil839, this PR is ready for another round of review. I have made the suggested changes.

Below are screenshots of the UI changes. Should I make the dropdown and input any longer, or are they fine?

<details>
<summary>Screenshots</summary>

![image](https://github.com/zulip/zulip/assets/87542880/1ea85267-8bd3-41ca-9806-3da61e10a7f7)

![image](https://github.com/zulip/zulip/assets/87542880/a7eb3bd9-7c03-4c99-8cd7-37f1f43ac344)
</details>

HemantUmre12 commented on web/styles/settings.css:
```diff
@@ -1457,6 +1458,22 @@ $option_title_width: 180px;
 
+#realm_jitsi_server_setting {
+    display: flex;
+    flex-direction: row;
```
Replaced `flex-direction` with `flex-wrap`.

sahil839 commented on web/src/settings_org.js:
```diff
@@ -1142,6 +1222,20 @@ function enable_or_disable_save_button($subsection_elem) {
+    if (time_limit_settings.length) {
+        disable_save_btn = should_disable_save_button_for_time_limit_settings(time_limit_settings);
+    } else if (is_video_chat_provider_jitsi_meet()) {
+        disable_save_btn = should_disable_save_button_for_jitsi_server_url_setting();
```
This should be called only for the relevant subsection. Currently this disables the save button for other subsections as well.

sahil839 commented on web/src/settings_org.js:
```diff
@@ -1109,8 +1174,23 @@ function check_maximum_valid_value($custom_input_elem, property_name) {
 
-function enable_or_disable_save_button($subsection_elem) {
-    const time_limit_settings = [...$subsection_elem.find(".time-limit-setting")];
+function is_valid_url(string) {
```
A better variable here would be `jitsi_url` and in case we decide to move this functions to a common file where it would be used by other parts of the code, then `url` would be better than string I guess.

sahil839 replied:
Another way to validate URL would be to use a regex but I am not sure which method will be efficient as we would check the validity of the input everytime the text in the input is changed.

sahil839 commented on web/src/settings_org.js:
```diff
@@ -258,9 +258,66 @@ function set_realm_waiting_period_setting() {
+
+function get_jitsi_server_url_setting_value($input_elem, for_api_data = true) {
+    if (!is_video_chat_provider_jitsi_meet()) {
+        return page_params.realm_jitsi_server_url;
```
This feels odd since we are passing the realm_jitsi_server_url even when video provider setting is not set to Jitsi. Would be better to add a comment explaining that we do this to make sure that the save button is hidden and API data does not contain the jitsi url parameter.

sahil839 reviewed (commented):
Posted some comments. Looks good overall. I am not sure what is the best UI as URLs can get long, but the current one seems fine. Maybe we can set the dropdown width to auto so it resizes itself as per the server-level setting, with a `min-width` set to make sure it is not too short

sahil839 commented:
We might also want to keep in mind how to handle the UI for short screens.

HemantUmre12 commented:
@sahil839, I have updated the PR. Please take a look.

Changes:

- Implemented regex for validating URLs in the frontend. Currently, we only validate URLs that start with `https://` or `http://`.

- Added `text-overflow: ellipsis` for the dropdown select to indicate when the URL is not fully visible. However, the complete URL will be visible in the dropdown options. If select width is set to `auto`, we will get horizontal scrolling on smaller screens when URLs are excessively long. Let me know if you have any suggestions for further improvements.

![image](https://github.com/zulip/zulip/assets/87542880/21ca043f-2580-40cb-9ccd-6d7551a2743f)

![image](https://github.com/zulip/zulip/assets/87542880/1fc1af03-807e-400b-8298-42daa775c5ee)

sahil839 commented on web/src/settings_org.js:
```diff
@@ -686,6 +747,9 @@ export function discard_property_element_changes(elem, for_realm_default_setting
             set_msg_move_limit_setting(property_name);
             break;
+        case "realm_jitsi_server_url":
+            set_video_chat_provider_dropdown();
```
Can't we call `set_jitsi_server_url_dropdown` here?

HemantUmre12 replied:
Considering that the Jitsi URL setting is dependent on the video chat provider, to prevent bugs when resetting the Jitsi URL setting to its initial state upon clicking "Discard," we need to first reset the video chat provider dropdown. Using `set_jitsi_server_url_dropdown` or skipping the case would work just fine, given that `set_video_chat_provider_dropdown` will be called ahead in the code path for the video chat provider setting, thereby resetting both settings again.
```
case "realm_jitsi_server_url":
    set_jitsi_server_url_dropdown();
    break;
```
```
// skipping
case "realm_jitsi_server_url":
    break;
```

I was thinking, using `set_video_chat_provider_dropdown` here would make it self-sufficient. The alternative mentioned above would also work just fine. What do you think?

sahil839 replied:
Reading the code now, I am not sure why the code for calling `set_video_chat_provider_dropdown();` is in `update_dependent_subsettings`. I think `update_dependent_subsettings` is to update the elements which are not directly for the changed setting but to update the elements dependent on changed setting, which does not seem to be the case here.

I think the correct code here should be that the `case` block for `realm_video_chat_provider` should be in discard_property_element_changes` itself and then after that call you can add `case` block for `realm_jitsi_server_url` which would only call `set_jitsi_server_url_dropdown`.

HemantUmre12 replied:
I'm not sure if this change should be a separate commit. For now, I've included the change in the same frontend commit.

sahil839 replied:
Would be good to have this in a separate commit. This commit already has many changes in `settings_org.js`.

HemantUmre12 replied:
@sahil839 done.

sahil839 reviewed (commented):
Posted a couple of comments.

I am really not sure about how to handle the long URLs case and how we want the UI to be. We have a couple of URL type inputs set to a width around 311px but we can't do this if we want both select and dropdown on the same line. I think we can leave the UI part as it is for now and update it after feedback from others, but I do think the simplest solution would be to just have both dropdown and custom input on different width and set the width similar to what we have for other url inputs.

HemantUmre12 commented:
@sahil839, I have addressed the comments above.

I'm also not too sure about the design. Moving forward, I believe we should finalize the design before making any further changes. For now, as suggested, I have set the width of both the select and input elements to be consistent with other URL inputs.

`width: 311px`
![image](https://github.com/zulip/zulip/assets/87542880/bf95392d-dfc7-4186-b79b-ad5fdcc93bd3)

sahil839 commented on web/styles/settings.css:
```diff
@@ -1459,6 +1460,26 @@ $option_title_width: 180px;
+    gap: 15px;
+
+    select {
+        min-width: auto;
```
This `min-width` is eventually computed as `0px` by the browser, so I don't think this is of any use now except that it prevents the dropdown to have width of `325px` which is the `min-width` for all other `select` elements.

Since, we are not sure about the widths currently, I think you can leave it as it is for now, but we should make sure to change it if needed after the UI is finalised to make sure we do not have any unnecessary or weird CSS rules.

timabbott replied:
Maybe add a comment "TODO: This is needed to override the 325px default for dropdowns but ideally we'd remove it." except maybe a bit clearer since I don't actually know what it's for -- but something to help a future developer know to clean it up.

HemantUmre12 replied:
Made some changes to the UI and now this is no longer required.

HemantUmre12 commented on web/src/compose.js:
```diff
@@ -89,8 +89,8 @@ export function update_video_chat_button_display() {
-        available_providers.jitsi_meet &&
-        page_params.realm_video_chat_provider === available_providers.jitsi_meet.id
+        page_params.realm_video_chat_provider === available_providers.jitsi_meet.id &&
+        page_params.jitsi_server_url
```
@sahil, I have added this commit. It is related to the changes made in #25922, which were merged a couple of days back. It was missing an edge case, and we make changes around here in the following commit. Therefore, I have included this commit in this PR. I don't know if removing `available_providers.jitsi_meet` was really necessary, but we also don't check it in `compute_show_video_chat_button()`.

sahil839 replied:
Ok, I also think that we do not need `available_providers.jitsi_meet` check.

sahil839 commented:
I have no other concerns. There are a couple of things that would need feedback from @timabbott like [this](https://github.com/zulip/zulip/pull/26399#discussion_r1288363272) and [this](https://github.com/zulip/zulip/pull/26399#discussion_r1288368872). Also for UI, we can probably start a discussion on #design to get feedback from more people.

timabbott commented on web/src/compose.js:
```diff
@@ -89,7 +89,7 @@ export function update_video_chat_button_display() {
     const available_providers = page_params.realm_available_video_chat_providers;
     if (
-        (available_providers.jitsi_meet &&
+        (page_params.jitsi_server_url &&
```
Would it make sense to leave this as `available_providers.jitsi_meet && page_params.jitsi_server_url`? I know `available_providers.jitsi_meet` is always true, but it seems more logically correct that way.

timabbott commented on zerver/models.py:
```diff
@@ -671,6 +671,9 @@ class Realm(models.Model):  # type: ignore[django-manager-missing] # django-stub
     )
 
+    JITSI_SERVER_SPECIAL_VALUES_MAP = {"default": None}
+    jitsi_server_url = models.TextField(null=True, default=None)
```
I think this should be using https://docs.djangoproject.com/en/4.2/ref/models/fields/#urlfield?

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -13437,6 +13448,19 @@ paths:
+                          - null, the default, means the realm will inherit the server-level setting.
+
+                          **Changes**: New in Zulip 8.0 (feature level 210). Previously, this was only
+                          available as a server-level configuration.
```
I think this description might be wrong; we have a separate/existing `jitsi_server_url` field (no `realm_` prefix) which is what the client should use to generate a call -- that value is always a URL (It will be the server-level value if this is unset.).

I think the correct description for this is:

> The URL of the custom Jitsi Meet server configured in this organization's settings. Clients should
> NOT use this field when starting video calls; the `jitsi_server_url` property should be used for that.
> This field is only used for the settings UI.
> 
> - `null`, the default, means this realm is configured to use the server's default Jitsi Meet server.

timabbott replied:
Started https://chat.zulip.org/#narrow/stream/378-api-design/topic/realm_jitsi_server_url/near/1637106 to try to figure out what we want to do here; will likely require some minor code changes too once we figure it out.

timabbott commented on zerver/tests/test_events.py:
```diff
@@ -2898,6 +2898,7 @@ def do_set_realm_property_test(self, name: str) -> None:
             video_chat_provider=[
                 Realm.VIDEO_CHAT_PROVIDERS["jitsi_meet"]["id"],
             ],
+            jitsi_server_url=["https://example1.jit.si", "https://example2.jit.si"],
```
Let's do `jitsi1.example.com`, and `jitsi2.example.com` -- we want to always use URLs in the official RFC `example.com` domain, not those of third parties, for this sort of thing.

timabbott commented on zerver/views/video_calls.py:
```diff
@@ -47,6 +47,15 @@ def __init__(self) -> None:
+    if value in special_values_map:
+        return special_values_map[value]
+
+    return value
```
Is there a reason for this to be in this module? I would think it should just sit above the function that uses it.

HemantUmre12 replied:
I don't exactly remember, maybe I confused the module with `lib/video_calls.py`. Moving it sounds good to me. I have updated the PR accordingly.

timabbott commented on zerver/views/realm.py:
```diff
@@ -276,6 +286,25 @@ def update_realm(
+                acting_user=user_profile,
+            )
+
+            data["jitsi_server_url"] = jitsi_server_url
```
This comment probably deserves some extra description or maybe an `else: pass` section -- I want it to be super clear that we expect `do_set_realm_property` to be called in the other cases, just via the loop below,

timabbott commented on zerver/tests/test_realm.py:
```diff
@@ -864,6 +864,36 @@ def test_message_retention_days(self) -> None:
+
+        req = dict(jitsi_server_url=orjson.dumps(12).decode())
+        result = self.client_patch("/json/realm", req)
+        self.assert_json_error(result, "jitsi_server_url is not an allowed_type")
```
This error message is quite awkward, but I think the right plan for deleting it is for @PIG208 to just migrate this file to the new `@typed_endpoint` system after we merge this.

timabbott commented on web/src/compose.js:
```diff
@@ -67,7 +67,8 @@ export function compute_show_video_chat_button() {
         page_params.realm_video_chat_provider === available_providers.jitsi_meet.id &&
-        !page_params.jitsi_server_url
+        !page_params.jitsi_server_url &&
+        !page_params.realm_jitsi_server_url
```
It's fairly awkward that we need to check both of these values.. but it is making me think we have the API not quite right.

timabbott commented on web/styles/settings.css:
```diff
@@ -1504,6 +1505,26 @@ $option_title_width: 180px;
+
+    & input,
+    select {
+        width: 311px;
```
Where does this 311 come from? A `calc()` or comment or CSS variable would usually be better.

HemantUmre12 replied:
I did some refactoring with the style, using the existing class `setting_url_input`, so we no longer need to define width here. I also made some design changes; I think labels on a separate line look better. Below are the before and after screenshots.

**Before:**

![](https://user-images.githubusercontent.com/87542880/261873935-bf95392d-dfc7-4186-b79b-ad5fdcc93bd3.png)

**After:**

![Screenshot 2023-09-20 001026](https://github.com/zulip/zulip/assets/87542880/a5f591e3-3167-49ed-a42c-0e90082bcf47)

![Screenshot 2023-09-20 001051](https://github.com/zulip/zulip/assets/87542880/420fcd75-4d37-4a79-bce7-ee43dbc49351)

timabbott commented on web/templates/settings/organization_settings_admin.hbs:
```diff
@@ -114,6 +114,30 @@
+                            </label>
+                            <select name="realm_jitsi_server_url" id="id_realm_jitsi_server_url" class="setting-widget prop-element settings_select bootstrap-focus-style" data-setting-widget-type="jitsi-server-url-setting">
+                                {{#if jitsi_server_url}}
+                                    <option value="server_default">{{ jitsi_server_url }}</option>
```
Can we have this say `https://foo.bar (default)`, to make it clear this is the special default value.

HemantUmre12 commented:
@timabbott The PR is ready for another round of review.

timabbott commented:
I rewrote the API documentation in a way that is hopefully helpful:

``` diff
diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
index 2aae48b9db..e9cb91cbb3 100644
--- a/zerver/openapi/zulip.yaml
+++ b/zerver/openapi/zulip.yaml
@@ -4370,15 +4370,14 @@ paths:
                                       nullable: true
                                       description: |
                                         The URL of the custom Jitsi Meet server configured in this organization's
-                                        settings. Clients should NOT use this field when starting video calls;
-                                        the `jitsi_server_url` property should be used for that.
-                                        This field is only used for the settings UI.
+                                        settings.
 
-                                        - null, the default, means this realm is configured to use the server's
-                                          default Jitsi Meet server.
+                                        `null`, the default, means that the organization is using the should use the
+                                        server-level configuration, `server_jitsi_server_url`.
 
                                         **Changes**: New in Zulip 8.0 (feature level 212). Previously, this was only
-                                        available as a server-level configuration.
+                                        available as a server-level configuration, and required a server restart to
+                                        change.
                                     waiting_period_threshold:
                                       type: integer
                                       description: |
@@ -13572,15 +13571,16 @@ paths:
                         nullable: true
                         description: |
                           The URL of the custom Jitsi Meet server configured in this organization's
-                          settings. Clients should NOT use this field when starting video calls;
-                          the `jitsi_server_url` property should be used for that.
-                          This field is only used for the settings UI.
+                          settings.
 
-                          - null, the default, means this realm is configured to use the server's
-                            default Jitsi Meet server.
+                          `null`, the default, means that the organization is using the should use the
+                          server-level configuration, `server_jitsi_server_url`. A correct client
+                          supporting only the modern API should use `realm_jitsi_server_url ||
+                          server_jitsi_server_url` to create calls.
 
                           **Changes**: New in Zulip 8.0 (feature level 212). Previously, this was only
-                          available as a server-level configuration.
+                          available as a server-level configuration, which was available via the
+                          `jitsi_server_url` field.
                       realm_giphy_rating:
                         type: integer
                         description: |
@@ -14037,10 +14037,20 @@ paths:
                                 on the external site.
                       jitsi_server_url:
                         type: string
+                        deprecated: true
                         description: |
                           Present if `realm` is present in `fetch_event_types`.
 
-                          The base URL the organization uses to create Jitsi video calls.
+                          The base URL to be used to create Jitsi video calls. Equals
+                          `realm_jitsi_server_url || server_jitsi_server_url`.
+
+                          **Changes**: Deprecated in Zulip 8.0 (feature level 212) and will
+                          eventually be removed. Previously, the Jitsi server to use was not
+                          configurable on a per-realm basis, and this field contained the server's
+                          configured Jitsi server. (Which is now provided as
+                          `server_jitsi_server_url`). Clients supporting older versions should fall
+                          back to this field when creating calls, e.g. `realm_jitsi_server_url ||
+                          server_jitsi_server_url || jitsi_server_url`.
                       development_environment:
                         type: boolean
                         description: |
@@ -14196,12 +14206,12 @@ paths:
                         type: string
                         nullable: true
                         description: |
-                          The URL of the Jitsi server that the Zulip server is configured to use.
-                          Clients should NOT use this field when starting video calls; the `jitsi_server_url`
-                          property should be used for that. This property is typically used in the settings
-                          UI to display the current server default.
+                          The URL of the Jitsi server that the Zulip server is configured to use by
+                          default; the organization-level setting `realm_jitsi_server_url` takes
+                          precendence over this setting when both are set.
 
-                          **Changes**: New in Zulip 8.0 (feature level 212).
+                          **Changes**: New in Zulip 8.0 (feature level 212). Previously, this value
+                          was available as the now-deprecated `jitsi_server_url`.
                       event_queue_longpoll_timeout_seconds:
                         type: integer
                         description: |
```

timabbott commented:
Overall, this looks very solid; I posted a comment just above with documentation changes, but otherwise I've marked this to merge once CI passes. Thanks so much for doing this @HemantUmre12 and @ligmitz!

timabbott commented:
Fixed a typo caught by `codespell`.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -14009,10 +14037,21 @@ paths:
+                          eventually be removed. Previously, the Jitsi server to use was not
+                          configurable on a per-realm basis, and this field contained the server's
+                          configured Jitsi server. (Which is now provided as
+                          `server_jitsi_server_url`). Clients supporting older versions should fall
+                          back to this field when creating calls: using `realm_jitsi_server_url ||
+                          server_jitsi_server_url` with newer servers and using `jitsi_server_url`
+                          with servers below feature level 212.
```
This can also be written as a double fallback, like `realm_jitsi_server_url ?? server_jitsi_server_url ?? jitsi_server_url`.

In particular that's how I think I'd want to write it in client code. Typically I've preferred a fallback over a version check, where possible; the code feels more robust that way, less sensitive to the exact details in the API docs, and also more transparent to the reader without having to go look up the API docs.

timabbott replied:
No, I meant what I wrote: The double fallback is I think wrong -- in the unlikely corner case of a modern server that starts out with `server_jitsi_server_url=None` and `realm_jitsi_server_url` set to `jitsi.example.com`, and one changes `realm_jitsi_server_url` to None, then `jitsi_server_url` will still have `jitsi.example.com`.

gnprice replied:
I see. I think I wasn't considering the possibility of the server-level value being null while the realm-level value isn't.

gnprice commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,17 @@ format used by the Zulip server that they are interacting with.
+  allowing organizations to set a custom Jitsi Meet server. Previously, this
+  was only available as a server-level configuration.
+
+* [`POST /register`](/api/register-queue): Added `server_jitsi_server_url`
+  fields to the `realm` object. The existing `jitsi_server_url` will now be
```
There's no `realm` object here, right? I believe what this wants to say is that fields `server_jitsi_server_url` and `realm_jitsi_server_url` were added.

timabbott replied:
Yeah, correct, those are added to the root, and of course when we extract `realm` and `server` objects, they'll land in different ones.

gnprice reviewed (commented):
A few comments on the docs.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -13553,6 +13566,21 @@ paths:
+
+                          `null`, the default, means that the organization is using the should use the
+                          server-level configuration, `server_jitsi_server_url`. A correct client
+                          supporting only the modern API should use `realm_jitsi_server_url ||
+                          server_jitsi_server_url` to create calls.
```
Yeah, this is meant to be pseudocode, but changing to the JS thing seems fine.

laurynmm commented:
Better late than never, made the revisions based on @gnprice's review above in #40033. All of it was still applicable.
